openstudiocore/src/model/EnergyManagementSystemProgram.cpp

```diff
@@ -0,0 +1,430 @@
+    std::string body_minus_comment;
+    std::vector<std::string> comments;
+    std::string comment;
+
+    //add program lines to body
+    std::vector<bool> ok(body_minus_nl.size(), false);
+    for (size_t i = 0; i < body_minus_nl.size(); i++) {
+      //split string on comment character !
+      comments = splitString(body_minus_nl.at(i), '!');
+      //get extensibleGroup object
+      WorkspaceExtensibleGroup group = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+      //make sure program line exists and insert
+      if (comments.size() > 0) {
+        //remove whitespace at end of comments[0]
+        boost::trim_right(comments[0]);
+        //remove whitespace at beginning of comments[0]
+        boost::trim_left(comments[0]);
+        //remove ,
+        pos = 0;
+        while ((pos = comments[0].find(",", pos)) != std::string::npos) {
+          comments[0].erase(pos, 1);
+        }
+        //insert program line
+        result = group.setString(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, comments[0]);
+        //check if comments exist
+        if (comments.size() > 1) {
+          //clear out the old comment
+          comment.clear();
+          //add comment to comment vector
+          comment.append(comments.at(1));
+          //combine any remaining comments for this line into one comment 
+          if (comments.size() > 2) {
+            for (size_t j = 2; j < comments.size(); j++) {
+              comment.append("!" + comments.at(j));
+            }
+          }
+          comment_result = group.setFieldComment(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, comment);
+        }
+      } else {
+        result = false;
+      }
+      ok.at(i) = result;
+    }
+    //check if all the programs set true
+    result = true;
+    for (size_t i = 0; i<ok.size(); i++) {
+      if (!ok.at(i)) {//the value is false
+        result = false; //not all values in array are true
+      }
+    }
+    return result;
+  }
+
+  bool EnergyManagementSystemProgram_Impl::eraseBody() {
+    //erase body of program
+    bool results = false;
+    std::vector< std::vector<std::string> > result;
+    result = clearExtensibleGroups();
+    // if result is not empty, then the group was cleared out since the old values were returned
+    if (!result.empty()) {
+      results = true;
+    };
+    return results;
+  }
+
+  bool EnergyManagementSystemProgram_Impl::addLine(const std::string& line) {
+    //add line to end of program body
+    bool result = true;
+    bool comment_result = false;
+    std::string comment;
+
+    // remove '\r' from the line string
+    std::string line_rn = line;
+    std::string::size_type pos = 0; // Must initialize
+    while ((pos = line_rn.find("\r", pos)) != std::string::npos) {
+      line_rn.erase(pos, 1);
+    }
+    // remove '\n' 
+    pos = 0;
+    while ((pos = line_rn.find("\n", pos)) != std::string::npos) {
```

Comment:
was easier to code since this came up at different times.  I can try and combine this